crates/matrix-sdk/tests/integration/encryption/verification.rs

```diff
@@ -99,33 +185,91 @@ impl MockedServer {
-                                .insert(key_id.to_owned(), signature.to_owned());
+                let mut keys = k.lock().unwrap();
+
+                for (p_user, p_signed_keys) in params.0 {
+                    for (p_key_id, p_json_diff) in p_signed_keys.iter() {
+                        // Try to find a field in keys.master.
+                        if let Some(existing_master_key) = keys.master.get_mut(&p_user) {
+                            let mut existing = existing_master_key.deserialize().unwrap();
+
+                            let target = DeviceKeyId::from_parts(
+                                ruma::DeviceKeyAlgorithm::Ed25519,
+                                p_key_id.into(),
+                            );
+
+                            if existing.keys.contains_key(&target) {
+                                let param: CrossSigningKey =
+                                    serde_json::from_str(p_json_diff.get()).unwrap();
+
+                                for (uid, sigs) in existing.signatures.iter_mut() {
+                                    if let Some(new_sigs) = param.signatures.get(uid) {
+                                        sigs.extend(new_sigs.clone());
+                                    }
+                                }
+                                for (uid, sigs) in param.signatures.iter() {
+                                    if !existing.signatures.contains_key(uid) {
+                                        existing.signatures.insert(uid.clone(), sigs.clone());
+                                    }
+                                }
+
+                                // Update in map.
+                                *existing_master_key = Raw::new(&existing).unwrap();
+                                continue;
+                            }
+                        }
+
+                        // Otherwise, try to find a field in keys.device.
+                        // Either merge signatures if entry is already present, or insert a new
+                        // entry.
+                        let known_devices = keys.device.entry(p_user.clone()).or_default();
+                        if let Some(device_keys) = known_devices.get_mut(p_key_id) {
+                            let p_json_diff: DeviceKeys =
+                                serde_json::from_str(p_json_diff.get()).unwrap();
+                            let mut deserialized_device_keys: DeviceKeys =
+                                serde_json::from_str(device_keys.get()).unwrap();
+                            deserialized_device_keys.signatures.extend(p_json_diff.signatures);
+                            *device_keys = RawValue::from_string(
+                                serde_json::to_string(&deserialized_device_keys).unwrap(),
+                            )
+                            .unwrap();
+                        } else {
+                            known_devices.insert(p_key_id.to_owned(), p_json_diff.to_owned());
```

Comment:
Hmm, I don't think that this is what the server would do, I think it would throw an error here. Is this necessary, I guess it might be because you might not be calling  `send_outgoing_requests()` which then means that some device might not have uploaded their device keys.

Reply:
The `sync()` calls `send_outgoing_requests()`, so we're fine, but see above comment: the signature upload could happen *before* the key was uploaded.